### PR TITLE
Initial sketch for a JS logging app

### PR DESCRIPTION
**Pull request by achamayou**
Follow up to #675, and first sub-item of #668, this is a minimal set of changes to run a JS version of the logging app against the same tests we use for other logging app implementations.

This is not:

1. a fully generic JS container, because:
  a. some more work is needed to expose the tables in a generic way, that ties into FlatBuffers, and probably also TypeScript types as tables schema.
  b. some more work is needed to avoid the json re-decoding/re-encoding, by allowing arbitrary payloads in/out (which is gated on killing off JSON-RPC)
2. sample code by any stretch of the imagination
3. the way we will package JS handlers of course, re-using the Lua stuff is expedient but not a permanent solution.

This does have support for console.log and exception traces however.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -365,6 +365,14 @@ if(BUILD_TESTS)
       CURL_CLIENT TRUE
     )
 
+    add_e2e_test(
+      NAME js_end_to_end_logging
+      PYTHON_SCRIPT ${CMAKE_SOURCE_DIR}/tests/e2e_logging.py
+      ADDITIONAL_ARGS
+        --js-app-script ${CMAKE_SOURCE_DIR}/src/apps/logging/loggingjs.lua
+      CURL_CLIENT TRUE
```
Comment (jumaffre) [inline CMakeLists.txt]: Minor: I don't think that the curl client is particularly useful here?

File: src/apps/jsgeneric/oe_sign.conf [hunk 2]
```diff
@@ -1,6 +1,6 @@
 # Enclave settings:
 Debug=1
-NumHeapPages=32768
+NumHeapPages=131072
```
Comment (jumaffre) [inline src/apps/jsgeneric/oe_sign.conf]: Why is this required?

**Comment by cimetrics** [comment]
qjs-run@[3932 aka 20200121.14](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3932&view=results) vs master ewma over 20 builds from [3767](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3767&view=results) to [3924](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3924&view=results)
images

File: CMakeLists.txt (hunk 1 above)
Comment (achamayou) [reply CMakeLists.txt]: Oh no, I agree, it's just because it's there in the other loggingenc one, but I can remove it.

File: src/apps/jsgeneric/oe_sign.conf (hunk 2 above)
Comment (achamayou) [reply src/apps/jsgeneric/oe_sign.conf]: The js impl seems to use a little more memory (I haven't checked how much, this is enough but probably too much).

File: src/apps/jsgeneric/jsgeneric.cpp [hunk 3]
```diff
@@ -13,33 +13,106 @@ namespace ccfapp
...
+        if (i != 0)
+          os << ' ';
+        str = JS_ToCString(ctx, argv[i]);
+        if (!str)
+          return JS_EXCEPTION;
+        os << str;
+        JS_FreeCString(ctx, str);
+      }
+      os << std::endl;
+      auto _ = logger::Out() == os;
```
Comment (eddyashton) [inline src/apps/jsgeneric/jsgeneric.cpp]: Do we need to deal with the internals of `logger` here? Or can we do something like:

```
std::stringstream ss;
for (i = 0; i < argc; ++i)
{
  if (i != 0) ss << ' ';
  str = JS_ToCString(ctx, argv[i]);
  if (!str) ...
  ss << str;
  JS_FreeCString(ctx, str);
}
LOG_INFO_FMT(ss.str());
```

File: src/apps/jsgeneric/jsgeneric.cpp [hunk 4]
```diff
@@ -61,63 +133,104 @@ namespace ccfapp
...
+        // TODO: share runtime across handlers?
+        // TODO: set memory limit with JS_SetMemoryLimit
+
+        JSContext* ctx = JS_NewContext(rt);
+        if (ctx == nullptr)
         {
-          int err_code = jsonrpc::CCFErrorCodes::SCRIPT_ERROR;
-          std::string msg = "";
+          JS_FreeRuntime(rt);
+          throw std::runtime_error("Failed to initialise QuickJS context");
```
Comment (eddyashton) [inline src/apps/jsgeneric/jsgeneric.cpp]: Should be "Failed to free".

File: src/apps/jsgeneric/jsgeneric.cpp [hunk 5]
```diff
@@ -13,33 +13,106 @@ namespace ccfapp
...
 
   class JSHandlers : public UserHandlerRegistry
   {
   private:
     NetworkTables& network;
+    LogTable& log_table;
 
   public:
-    JSHandlers(NetworkTables& network, const uint16_t n_tables = 8) :
+    JSHandlers(NetworkTables& network, const uint16_t n_tables = 0) :
```
Comment (eddyashton) [inline src/apps/jsgeneric/jsgeneric.cpp]: `n_tables` can be removed.

**Comment by cimetrics** [comment]
qjs-run@[3934 aka 20200121.15](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3934&view=results) vs master ewma over 20 builds from [3767](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3767&view=results) to [3924](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3924&view=results)
images

File: src/apps/jsgeneric/jsgeneric.cpp (hunk 4 above)
Comment (achamayou) [reply src/apps/jsgeneric/jsgeneric.cpp]: No? A parent runtime can have multiple contexts (each one has its own heap), and here we're freeing the (sole) parent runtime if we fail to create a context successfully, but the error is that we failed to initialise a context.

File: src/apps/jsgeneric/jsgeneric.cpp (hunk 5 above)
Comment (achamayou) [reply src/apps/jsgeneric/jsgeneric.cpp]: Yes, thank you.

File: src/apps/jsgeneric/jsgeneric.cpp (hunk 3 above)
Comment (achamayou) [reply src/apps/jsgeneric/jsgeneric.cpp]: Thanks, I didn't think of that!

Comment (achamayou) [reply src/apps/jsgeneric/jsgeneric.cpp]: Done

**Comment by cimetrics** [comment]
qjs-run@[3944 aka 20200121.19](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3944&view=results) vs master ewma over 20 builds from [3771](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3771&view=results) to [3940](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3940&view=results)
images

File: src/apps/jsgeneric/jsgeneric.cpp [hunk 6]
```diff
@@ -13,33 +13,101 @@ namespace ccfapp
   using namespace kv;
   using namespace ccf;
 
-  using GenericTable = ccf::Store::Map<nlohmann::json, nlohmann::json>;
+  using LogTable = ccf::Store::Map<int32_t, std::string>;
+
+  static JSValue js_print(
+    JSContext* ctx, JSValueConst this_val, int argc, JSValueConst* argv)
```
Comment (olgavrou) [inline src/apps/jsgeneric/jsgeneric.cpp]: `this_val` is unused

Comment (achamayou) [reply src/apps/jsgeneric/jsgeneric.cpp]: Yes, this is effectively a static call, it doesn't need access to the console object which is its "this". It still needs to be there in the signature for the call to work though.

File: src/apps/jsgeneric/jsgeneric.cpp [hunk 7]
```diff
@@ -13,33 +13,101 @@ namespace ccfapp
...
+  {
+    JSValue exception_val = JS_GetException(ctx);
+
+    JSValue val;
+    const char* stack;
+    bool is_error;
+
+    is_error = JS_IsError(ctx, exception_val);
+    if (!is_error)
+      LOG_INFO_FMT("Throw: ");
```
Comment (jumaffre) [inline src/apps/jsgeneric/jsgeneric.cpp]: Minor: should these be `LOG_FAIL_FMT()`?

Comment (achamayou) [reply src/apps/jsgeneric/jsgeneric.cpp]: No, FAIL is an important failure inside CCF, this is just business logic raising an exception. It's perfectly safe, handled etc. In fact, in production code, this should be turned off, because it can leak information about what the business logic is doing.

File: src/apps/logging/loggingjs.lua [hunk 8]
```diff
@@ -0,0 +1,38 @@
+-- Copyright (c) Microsoft Corporation. All rights reserved.
+-- Licensed under the Apache 2.0 License.
+
+return {
```
Comment (jumaffre) [inline src/apps/logging/loggingjs.lua]: Should this file have the `.lua` extension?

File: src/apps/jsgeneric/jsgeneric.cpp [hunk 9]
```diff
@@ -13,33 +13,101 @@ namespace ccfapp
...
+
+  static JSValue js_put(
+    JSContext* ctx, JSValueConst this_val, int argc, JSValueConst* argv)
+  {
+    auto log_table_view = (LogTable::TxView*)JS_GetContextOpaque(ctx);
+    if (!JS_IsInteger(argv[0]))
+      return JS_EXCEPTION;
+    int32_t i = JS_VALUE_GET_INT(argv[0]);
+    auto v = JS_ToCString(ctx, argv[1]); // TODO: error checking?
+    std::string s(v);
```
Comment (jumaffre) [inline src/apps/jsgeneric/jsgeneric.cpp]: Could we use a `string_view` here to avoid the copy?

Comment (achamayou) [reply src/apps/jsgeneric/jsgeneric.cpp]: I think we can skip the intermediary altogether, yes.

**Comment by cimetrics** [comment]
qjs-run@[3964 aka 20200122.4](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3964&view=results) vs master ewma over 20 builds from [3772](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3772&view=results) to [3958](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3958&view=results)
images

File: src/apps/logging/loggingjs.lua (hunk 8 above)
Comment (achamayou) [reply src/apps/logging/loggingjs.lua]: Yes, it's lua! See point 3. in PR description, this lets us re-use existing infrastructure for now.
